src/preprocessing/target_encoders.rs

```diff
@@ -0,0 +1,233 @@
+#![allow(clippy::ptr_arg)]
+//! # Encode categorical features as a one-hot or multi-class numeric array.
+
+use crate::error::Failed;
+use crate::math::num::RealNumber;
+use std::collections::HashMap;
+use std::hash::Hash;
+
+/// Turn a collection of `LabelType`s into a one-hot vectors.
+/// This struct encodes single class per exmample
+///
+/// You can fit a label enumeration by passing a collection of labels.
+/// Label numbers will be assigned in the order they are encountered
+///
+/// Example:
+/// ```
+/// use std::collections::HashMap;
+/// use smartcore::preprocessing::target_encoders::OneHotEncoder;
+///
+/// let fake_labels: Vec<usize> = vec![1,2,3,4,5,3,5,3,1,2,4];
+/// let enc = OneHotEncoder::<usize>::fit(&fake_labels[..]);
+/// let oh_vec: Vec<f64> = enc.transform_one(&1).unwrap();
+/// // notice that 1 is actually a zero-th positional label
+/// assert_eq!(oh_vec, vec![1.0, 0.0, 0.0, 0.0, 0.0]);
+/// ```
+///
+/// You can also pass a predefined label enumeration such as a hashmap `HashMap<LabelType, usize>` or a vector `Vec<LabelType>`
+///
+///
+/// ```
+/// use std::collections::HashMap;
+/// use smartcore::preprocessing::target_encoders::OneHotEncoder;
+///
+/// let label_map: HashMap<&str, usize> =
+/// vec![("cat", 2), ("background",0), ("dog", 1)]
+/// .into_iter()
+/// .collect();
+/// let label_vec = vec!["background", "dog", "cat"];
+///
+/// let enc_lv = OneHotEncoder::<&str>::from_positional_label_vec(label_vec);
+/// let enc_lm = OneHotEncoder::<&str>::from_label_map(label_map);
+///
+/// // ["background", "dog", "cat"]
+/// println!("{:?}", enc_lv.get_labels());
+/// assert_eq!(enc_lv.transform_one::<f64>(&"dog"), enc_lm.transform_one::<f64>(&"dog"))
+/// ```
+pub struct OneHotEncoder<LabelType> {
+    label_to_idx: HashMap<LabelType, usize>,
+    labels: Vec<LabelType>,
+    num_classes: usize,
+}
+
+enum LabelDefinition<T> {
+    LabelToClsNumMap(HashMap<T, usize>),
+    PositionalLabel(Vec<T>),
+}
+
+/// Crearte a vector of size num_labels with zeros everywhere and 1 at label_idx (one-hot vector)
+pub fn make_one_hot<T: RealNumber>(label_idx: usize, num_labels: usize) -> Vec<T> {
+    let (pos, neg) = (T::from_f64(1f64).unwrap(), T::from_f64(0f64).unwrap());
+    (0..num_labels)
+        .map(|idx| if idx == label_idx { pos } else { neg })
+        .collect()
+}
+
+impl<'a, LabelType: Hash + Eq + Clone> OneHotEncoder<LabelType> {
+    /// Fit an encoder to a lable list
+    pub fn fit(labels: &[LabelType]) -> Self {
```

Comment:
Can you please rename this method? Method `fit` is reserved for an operation that estimates parameters of a transformer from n-dimensional array.  Here is a good example of how method `fit` should look like https://github.com/smartcorelib/smartcore/blob/development/src/decomposition/pca.rs#L119